### PR TITLE
grpc: add a raw (Buffer::Instance) interface

### PR DESCRIPTION
**Pull request by jplevyak**
For an explanation of how to fill out the fields, please see the relevant section 
in [PULL_REQUESTS.md](./PULL_REQUESTS.md)

Description: Add a raw (Buffer::Instance) gRPC interface.  The existing interface is layered on top
of a new raw interface which takes the service and method names as absl::string_view and which
passes the data to be sent and data received as Buffer::Instance(s).  This is similar to the way
the existing typed interface is layered on the untyped interface.  This enables extensions and
plugins to use gRPC without the gRPC service protobufs being compiled into envoy.  This can also
support general gRPC proxies. 
Risk Level: Medium
Testing: all existing tests pass.
Docs Changes: N/A (no externally visible change)
Release Notes: N/A
[Optional Fixes #Issue]
[Optional Deprecated:]

### CODE REVIEW

File: include/envoy/grpc/async_client.h [hunk 1]
```diff
@@ -41,7 +42,16 @@ class AsyncStream {
...
+  virtual void sendMessage(const Protobuf::Message& request, bool end_stream);
+
+  /**
+   * Send request message to the stream.
+   * @param request serializalized message.
+   * @param end_stream close the stream locally. No further methods may be invoked on the stream
+   *                   object, but callbacks may still be received until the stream is closed
+   *                   remotely.
+   */
+  virtual void sendRawMessage(Buffer::InstancePtr request, bool end_stream) PURE;
```
Comment (lizan) [inline include/envoy/grpc/async_client.h]: does the first parameter has to be an unique_ptr? seems `Buffer::Instance&` should be fine?

File: include/envoy/grpc/async_client.h [hunk 2]
```diff
@@ -71,15 +81,18 @@ class AsyncRequestCallbacks {
    * Factory for empty response messages.
    * @return ProtobufTypes::MessagePtr a Protobuf::Message with the response
    *         type for the request.
+   * NB: createEmptyResponse will not be called if onSuccessRaw() is overriden.
    */
-  virtual ProtobufTypes::MessagePtr createEmptyResponse() PURE;
+  virtual ProtobufTypes::MessagePtr createEmptyResponse() {
+    throw EnvoyException("AsyncRequestCallbacks::createEmptyResponse must be overriden");
```
Comment (lizan) [inline include/envoy/grpc/async_client.h]: NOT_REACHED?

File: include/envoy/grpc/async_client.h [hunk 3]
```diff
@@ -108,20 +135,23 @@ template <class ResponseType> class TypedAsyncRequestCallbacks : public AsyncReq
...
 
   /**
    * Factory for empty response messages.
    * @return ProtobufTypes::MessagePtr a Protobuf::Message with the response
    *          type for the stream.
+   * NB: createEmptyResponse will not be called if onRecieveRawMessage() is overriden.
    */
-  virtual ProtobufTypes::MessagePtr createEmptyResponse() PURE;
+  virtual ProtobufTypes::MessagePtr createEmptyResponse() {
+    throw EnvoyException("AsyncStreamCallbacks::createEmptyResponse must be overriden");
```
Comment (lizan) [inline include/envoy/grpc/async_client.h]: ditto

File: include/envoy/grpc/async_client.h [hunk 4]
```diff
@@ -91,6 +104,20 @@ class AsyncRequestCallbacks {
                          Tracing::Span& span) PURE;
 };
 
+class AsyncRequestCallbacks : public RawAsyncRequestCallbacks {
```
Comment (lizan) [inline include/envoy/grpc/async_client.h]: Do we really need this class? Can `RawAsyncRequestCallbacks` act as this?

File: source/common/buffer/buffer_impl.cc [hunk 5]
```diff
@@ -130,8 +130,10 @@ void OwnedImpl::commit(RawSlice* iovecs, uint64_t num_iovecs) {
       slice_index--;
     }
     if (slice_index < 0) {
-      // There was no slice containing any data, so rewind the iterator at the first slice.
       slice_index = 0;
+      if (!slices_[0]) {
```
Comment (lizan) [inline source/common/buffer/buffer_impl.cc]: why this change is needed?

File: source/common/grpc/async_client.cc [hunk 6]
```diff
@@ -0,0 +1,55 @@
+#include "envoy/grpc/async_client.h"
```
Comment (lizan) [inline source/common/grpc/async_client.cc]: I'm not keen to this structure, implementing interface in `include` without Impl classes. It also add an transitive dependency of the interface. 

Can we have Raw and Typed classes, but not inherit but just own Raw in Typed?

File: include/envoy/grpc/async_client.h (hunk 1 above)
Comment (jplevyak) [reply include/envoy/grpc/async_client.h]: The caller has no further need for the buffer and the callee may need to keep it (queue it) so it makes sense to eliminate the overhead of copying it and pass it as a unique ptr.   In other bits of the code this is done via && (e.g. Buffer::InstancePtr &&), but given that a std::unique_ptr is only 16 bytes that doesn't make much sense to me.

File: include/envoy/grpc/async_client.h [hunk 7]
```diff
@@ -91,6 +104,20 @@ class AsyncRequestCallbacks {
...
+public:
+  virtual ~AsyncRequestCallbacks() {}
+
+  void onSuccessRaw(Buffer::InstancePtr response, Tracing::Span& span) override;
```
Comment (htuch) [inline include/envoy/grpc/async_client.h]: Comments? Why do we have the same callback class providing both raw and untyped? Can't one be implemented in terms of the other?

File: include/envoy/grpc/async_client.h [hunk 8]
```diff
@@ -195,7 +247,24 @@ class AsyncClient {
...
+   * @param method_name name of the method (i.e. service_method.name()).
+   * @param request serialized message.
+   * @param callbacks the callbacks to be notified of RPC status.
+   * @param parent_span the current parent tracing context.
+   * @param timeout supplies the request timeout.
+   * @return a request handle or nullptr if no request could be started. NOTE: In this case
+   *         onFailure() has already been called inline. The client owns the request and the
+   *         handle should just be used to cancel.
+   */
+  virtual AsyncRequest* sendRaw(absl::string_view service_full_name, absl::string_view method_name,
```
Comment (htuch) [inline include/envoy/grpc/async_client.h]: Yeah, I reckon we should have the gRPC implementations just implement the raw variants, and we can layer a wrapper layer to do the rest on top of this. That will shrink the interface a bit.

File: include/envoy/grpc/async_client.h [hunk 9]
```diff
@@ -91,6 +104,20 @@ class AsyncRequestCallbacks {
...
+  virtual ~AsyncRequestCallbacks() {}
+
+  void onSuccessRaw(Buffer::InstancePtr response, Tracing::Span& span) override;
+  /**
+   * Called when the async gRPC request succeeds. No further callbacks will be invoked.
+   * @param response the gRPC response.
+   * @param span a tracing span to fill with extra tags.
+   * NB: requires overriding createEmptyResponse().
+   */
+  virtual void onSuccessUntyped(ProtobufTypes::MessagePtr&& response, Tracing::Span& span) PURE;
```
Comment (htuch) [inline include/envoy/grpc/async_client.h]: I think if we have in the "this is the interface for a gRPC implementation" only raw callbacks/methods, and then in `source/common/grpc` we can add an lightweight proto -> raw wrapper layer, this could simplify the interface.

File: include/envoy/grpc/async_client.h (hunk 1 above)
Comment (jplevyak) [reply include/envoy/grpc/async_client.h]: This is a generic interface, so it seems like "const Buffer::Instance&" makes the most sense if the copy is reasonably efficient.   What do you guys think?

File: include/envoy/grpc/async_client.h (hunk 2 above)
Comment (jplevyak) [reply include/envoy/grpc/async_client.h]: NOT_REACHED_GCOVR_EXCL_LINE it is.

File: include/envoy/grpc/async_client.h (hunk 7 above)
Comment (jplevyak) [reply include/envoy/grpc/async_client.h]: If you subclass from AsyncRequestCallbacks then it overrides onSuccessRaw for you so that onSuccessUntyped is called as you are expected to override that.

I can make onSuccessRaw private in AsyncRequestCallbacks to make that clear and prevent accidental overrides of the wrong function.

File: include/envoy/grpc/async_client.h (hunk 9 above)
Comment (jplevyak) [reply include/envoy/grpc/async_client.h]: I struggled a bit with the best way to do this, and decided to follow the example of *Untyped here.  I am not sure how your suggestion would work out for the common case since typed messages are likely to be used fairly commonly.

Perhaps we can talk offline and you can give me a better idea of what you are proposing here.

File: include/envoy/grpc/async_client.h (hunk 8 above)
Comment (jplevyak) [reply include/envoy/grpc/async_client.h]: Let's discuss this.  The implementation only works with the Raw version except for Protobuf::MethodDescriptor which it must convert to the service_full_name and method_name.  This conversion accounts for start/startRaw ->send/sendRaw.   For clarity the raw forms also take the raw versions of the callbacks.  However the implementations do not know about the non-Raw forms of the callbacks.  So in theory we could just overload send()/start() to do the Protobuf::MethodDescriptor conversion and have a single for for the callbacks, but then we would still want the other forms of callbacks although we could move them into the source/... tree.  However I am not sure how that would work with the client code and it would complicate the tests.  Right now we can mock all the versions and do the tests at the most natural level (typically not Raw).  Let's discuss it.

File: source/common/buffer/buffer_impl.cc (hunk 5 above)
Comment (jplevyak) [reply source/common/buffer/buffer_impl.cc]: The existing code can't send a fully empty OwnedImpl because slices_[0] is nullptr.  This just patches that up.  Prior to this change completely empty Buffer::Instances were created differently (e.g. out of std::string) which did not have this problem.

File: source/common/grpc/async_client.cc (hunk 6 above)
Comment (jplevyak) [reply source/common/grpc/async_client.cc]: Is this similar to what htuch@ is suggesting about having the Typed version in source/... ?  It seems like it would be possible to do what you are suggesting, although keeping the Typed version abstract at the interface level without inheritance could get tricky, e.g. requiring a factory for example which might be as complicated.   If this was just implementation then it is definitely possible.  However as I said above, it would make mocking a bit trickier and it would require dipping into the implementation in the common case.  Could give me a bit more of an idea of what you are proposing?

File: include/envoy/grpc/async_client.h (hunk 4 above)
Comment (jplevyak) [reply include/envoy/grpc/async_client.h]: This class provides support for onSuccessUntyped which in turn provides support for onSuccess (the typed version).

File: include/envoy/grpc/async_client.h (hunk 3 above)
Comment (jplevyak) [reply include/envoy/grpc/async_client.h]: done.

**Comment by mattklein123** [comment]
Can I ask a high level probably stupid question: how is this actually going to be used? Where are the raw message bytes going to come from to send and what is the caller going to do with the raw bytes returned? I understand this change in theory but the utility is not clear to me.

**Comment by jplevyak** [comment]
> Can I ask a high level probably stupid question: how is this actually going to be used? Where are the raw message bytes going to come from to send and what is the caller going to do with the raw bytes returned? I understand this change in theory but the utility is not clear to me.

This is going to be used by the WASM extension system.  It can also be used by any generic gRPC proxy system.

**Comment by jplevyak** [comment]
I believe I have a plan to address the issue of the API:
  - make the base API the raw API
    - startRaw -> start where the clients convert Protobuf::MethodDescriptor
  - move the Untyped and Typed Callback support code to source/common/grpc/grpc_callbacks.*
    - clients will use to provide the interface used today
  - continue having mocks for the Untyped and Typed Callbacks in the test/mocks directory

I'll ping this PR when I am done.

Thanx for the comments.

**Comment by mattklein123** [comment]
> This is going to be used by the WASM extension system. It can also be used by any generic gRPC proxy system.

OK, thanks that makes sense. So basically the WASM code will have the protos compiled in and just use a raw API to make calls. Regarding the latter, Envoy already is a generic gRPC proxy so I'm not sure why this would ever be used in that manner, but agree it seems possible. Since this is targeted towards WASM/plugin can you update the issue description to make that more clear?

Also, more generally, @jplevyak @PiotrSikora @lizan @dio is there a design doc on the proposed WASM <-> API bindings? I've been asking for this for a while and would love to review. Thank you.

File: include/envoy/grpc/async_client.h (hunk 1 above)
Comment (htuch) [reply include/envoy/grpc/async_client.h]: Slice movement is relatively efficient, but that's not copying. It doesn't have to be `const`, you can donate slices between two mutable buffers.

**Comment by htuch** [comment]
@jpleyvak this seems like a good plan. Please ping me on IM if you want to chat further. The key things to preserve are economy of mechanism and mockability :)

**Comment by jplevyak** [comment]
I have updated this PR as per the comments.  PTAL

File: source/common/grpc/common.h [hunk 10]
```diff
@@ -119,10 +120,15 @@ class Common {
                                       std::string* method);
 
   /**
-   * Serialize protobuf message.
+   * Serialize protobuf message. With grpc header.
```
Comment (lizan) [inline source/common/grpc/common.h]: nit: with grpc frame header.

How about call this method `serializeToGrpcFrame` and the one without grpc header `serializeMessage`? Having `serializeBody` change its semantic will be problematic for out-of-tree code.

File: test/common/grpc/common_test.cc [hunk 11]
```diff
@@ -349,5 +349,15 @@ TEST(GrpcCommonTest, ValidateResponse) {
   }
 }
 
+TEST(GrpcCommonTest, MakeBufferInstance) {
```
Comment (lizan) [inline test/common/grpc/common_test.cc]: more tests for this? e.g. buffer with a few slices, validating they are same. (same for below)

File: source/common/grpc/typed_async_client.h [hunk 12]
```diff
@@ -0,0 +1,145 @@
...
+  bool operator==(AsyncStream* stream) const { return stream_ == stream; }
+  bool operator!=(AsyncStream* stream) const { return stream_ != stream; }
+
+private:
+  template <typename Request> friend class TypedAsyncStream;
+  AsyncStream* stream_{};
+  bool is_grpc_header_required_{};
+};
+
+template <typename Request> class TypedAsyncStream : public UntypedAsyncStream {
```
Comment (htuch) [inline source/common/grpc/typed_async_client.h]: What is the benefit of having two classes here? The untyped one is actually just a proto wrapper around the real untyped async client (and is somewhat typed, it has a proto interface), then we have the typed wrapper which doesn't seem to do very much. It seems we could collapse the two together?

File: source/common/grpc/typed_async_client.h [hunk 13]
```diff
@@ -0,0 +1,145 @@
...
+public:
+  virtual ~TypedAsyncStreamCallbacks() {}
+  virtual void onReceiveMessageTyped(std::unique_ptr<Response>&& message) PURE;
+
+private:
+  ProtobufTypes::MessagePtr createEmptyResponse() override { return std::make_unique<Response>(); }
+  void onReceiveMessageUntyped(ProtobufTypes::MessagePtr&& response) override {
+    onReceiveMessageTyped(std::unique_ptr<Response>(dynamic_cast<Response*>(response.release())));
+  }
+};
```
Comment (htuch) [inline source/common/grpc/typed_async_client.h]: Same for all these dual untyped/typed wrappers.

File: source/common/grpc/typed_async_client.cc [hunk 14]
```diff
@@ -0,0 +1,63 @@
...
+#include "common/buffer/zero_copy_input_stream_impl.h"
+#include "common/common/utility.h"
+#include "common/grpc/common.h"
+#include "common/http/utility.h"
+
+namespace Envoy {
+namespace Grpc {
+
+void UntypedAsyncStream::sendMessageUntyped(const Protobuf::Message& request, bool end_stream) {
+  stream_->sendMessage(stream_->isGrpcHeaderRequired() ? Common::serializeBody(request)
```
Comment (htuch) [inline source/common/grpc/typed_async_client.cc]: We should discuss why this is needed. I know you left some IM about this, but it's disappeared, sorry about that. I would ideally not have any special case logic here.

File: source/common/grpc/common.h [hunk 15]
```diff
@@ -148,6 +154,21 @@ class Common {
    */
   static std::string typeUrl(const std::string& qualified_name);
 
+  /**
+   * BUild grpc::ByteBuffer which aliases the data in a Buffer::InstancePtr.
+   * @param bufferInstance source data container.
+   * @return byteBuffer target container aliased to the data in Buffer::Instance and owning the
+   * Buffer::Instance.
+   */
+  static grpc::ByteBuffer makeByteBuffer(Buffer::InstancePtr bufferInstance);
```
Comment (htuch) [inline source/common/grpc/common.h]: Can you split these into a separate PR (with their tests)? They are self contained helpers that would be easier to review in isolation.

File: include/envoy/grpc/async_client.h (hunk 1 above)
Comment (jplevyak) [reply include/envoy/grpc/async_client.h]: Ended up with Buffer::InstancePtr&& which seemed to be most consistent with other interfaces.

File: include/envoy/grpc/async_client.h (hunk 7 above)
Comment (jplevyak) [reply include/envoy/grpc/async_client.h]: Updated to raw only API

File: include/envoy/grpc/async_client.h (hunk 9 above)
Comment (jplevyak) [reply include/envoy/grpc/async_client.h]: Done.

File: include/envoy/grpc/async_client.h (hunk 4 above)
Comment (jplevyak) [reply include/envoy/grpc/async_client.h]: Done

File: include/envoy/grpc/async_client.h (hunk 8 above)
Comment (jplevyak) [reply include/envoy/grpc/async_client.h]: Updated to the raw only API.

File: source/common/grpc/async_client.cc (hunk 6 above)
Comment (jplevyak) [reply source/common/grpc/async_client.cc]: Updated to raw only API.

File: source/common/grpc/common.h (hunk 10 above)
Comment (jplevyak) [reply source/common/grpc/common.h]: Done

File: source/common/grpc/typed_async_client.cc (hunk 14 above)
Comment (jplevyak) [reply source/common/grpc/typed_async_client.cc]: This is needed because the wrappers need to serialize to the raw buffer form and that format is different depending whether or not grpc framing is required.  Yes, that was the subject of the IM; I'll try to turn on history so that things won't get lost in the future.

File: source/common/grpc/typed_async_client.h (hunk 12 above)
Comment (jplevyak) [reply source/common/grpc/typed_async_client.h]: This was the format that we had before.  I am happy to get rid of it, but I was trying to preserve the original behavior.

File: include/envoy/grpc/async_client.h [hunk 16]
```diff
@@ -55,6 +57,11 @@ class AsyncStream {
    * stream object and no further callbacks will be invoked.
    */
   virtual void resetStream() PURE;
+
+  /**
+   * @return true if the underlying protocol requires messages to be serialized with a gRPC header.
+   */
+  virtual bool isGrpcHeaderRequired() PURE;
```
Comment (lizan) [inline include/envoy/grpc/async_client.h]: `isGrpcFrameRequired`?

Comment (jplevyak) [reply include/envoy/grpc/async_client.h]: Removed this.  PTAL

File: source/common/grpc/common.h (hunk 15 above)
Comment (jplevyak) [reply source/common/grpc/common.h]: Let's get the API settled, then I will split these and follow up on this thread.

File: source/common/grpc/typed_async_client.cc (hunk 14 above)
Comment (jplevyak) [reply source/common/grpc/typed_async_client.cc]: As discussed, I removed isGrpcHeaderRequired().  PTAL

File: source/common/grpc/typed_async_client.h (hunk 13 above)
Comment (jplevyak) [reply source/common/grpc/typed_async_client.h]: Done.  PTAL

File: source/common/grpc/typed_async_client.h (hunk 12 above)
Comment (jplevyak) [reply source/common/grpc/typed_async_client.h]: Done. PTAL

File: test/common/grpc/common_test.cc (hunk 11 above)
Comment (jplevyak) [reply test/common/grpc/common_test.cc]: When we are good with the API I'll break out the common changes into a separate PR and add more tests.

**Comment by jplevyak** [comment]
I have created https://github.com/envoyproxy/envoy/pull/6732 which contains the utilities changes and the rename -> serializeToGrpcFrame from this PR.  PTAL.

**Comment by jplevyak** [comment]
Was out of town, back on this.

**Comment by jplevyak** [comment]
Updated to master.  What can we break off next?

File: test/common/grpc/common_test.cc (hunk 11 above)
Comment (jplevyak) [reply test/common/grpc/common_test.cc]: done

**Comment by htuch** [comment]
@jplevyak I think this is good to review in a single PR, I will give it another pass in the next day.

File: source/common/grpc/BUILD [hunk 17]
```diff
@@ -62,7 +76,10 @@ envoy_cc_library(
     name = "common_lib",
     srcs = ["common.cc"],
     hdrs = ["common.h"],
-    external_deps = ["abseil_optional"],
+    external_deps = [
+        "abseil_optional",
+        "grpc",
```
Comment (htuch) [inline source/common/grpc/BUILD]: Do we need to take a Google gRPC dependency here?

File: source/common/grpc/async_client_impl.cc [hunk 18]
```diff
@@ -129,21 +131,18 @@ void AsyncStreamImpl::onData(Buffer::Instance& data, bool end_stream) {
...
+                                                    : std::make_unique<Buffer::OwnedImpl>())) {
+      streamError(Status::GrpcStatus::Internal);
+      return;
     }
-    callbacks_.onReceiveMessageUntyped(std::move(response));
   }
 
-  if (end_stream) {
-    Http::HeaderMapPtr empty_trailers = std::make_unique<Http::HeaderMapImpl>();
+  if (!http_reset_ && end_stream) {
```
Comment (htuch) [inline source/common/grpc/async_client_impl.cc]: Can you explain this subtle change in logic?

File: source/common/grpc/typed_async_client.h [hunk 19]
```diff
@@ -0,0 +1,114 @@
...
+namespace Envoy {
+namespace Grpc {
+
+/**
+ * Forward declarations for helper functions.
+ */
+void SendMessageUntyped(RawAsyncStream* stream, const Protobuf::Message& request, bool end_stream);
+ProtobufTypes::MessagePtr ParseMessageUntyped(ProtobufTypes::MessagePtr&& message,
+                                              Buffer::InstancePtr&& response);
+RawAsyncStream* StartUntyped(RawAsyncClient* client,
```
Comment (htuch) [inline source/common/grpc/typed_async_client.h]: Nit: please lowerCase all method names in this PR.

File: source/common/grpc/typed_async_client.h [hunk 20]
```diff
@@ -0,0 +1,114 @@
...
+
+#include "envoy/grpc/async_client.h"
+
+namespace Envoy {
+namespace Grpc {
+
+/**
+ * Forward declarations for helper functions.
+ */
+void SendMessageUntyped(RawAsyncStream* stream, const Protobuf::Message& request, bool end_stream);
```
Comment (htuch) [inline source/common/grpc/typed_async_client.h]: Can we do something to indicate that these are internal? It's a bit confusing having untyped methods in a typed async client.

File: test/common/config/delta_subscription_test_harness.h [hunk 21]
```diff
@@ -88,10 +89,13 @@ class DeltaSubscriptionTestHarness : public SubscriptionTestHarness {
...
-        .WillOnce([this](const Protobuf::Message& message, bool) {
-          const std::string nonce =
-              static_cast<const envoy::api::v2::DeltaDiscoveryRequest&>(message).response_nonce();
+                sendMessageRaw_(
+                    Grpc::ProtoBufferEqIgnoringField(expected_request, "response_nonce"), false))
+        .WillOnce([this](Buffer::InstancePtr& buffer, bool) {
+          Buffer::ZeroCopyInputStreamImpl stream(std::move(buffer));
+          envoy::api::v2::DeltaDiscoveryRequest message;
+          EXPECT_TRUE(message.ParseFromZeroCopyStream(&stream));
+          const std::string nonce = message.response_nonce();
```
Comment (htuch) [inline test/common/config/delta_subscription_test_harness.h]: This pattern comes up a few times, can you refactor?

File: test/mocks/grpc/mocks.h [hunk 22]
```diff
@@ -91,5 +102,36 @@ class MockAsyncClientManager : public AsyncClientManager {
                                      Stats::Scope& scope, bool skip_cluster_check));
 };
 
+MATCHER_P(ProtoBufferEq, expected, "") {
```
Comment (htuch) [inline test/mocks/grpc/mocks.h]: Check out https://github.com/envoyproxy/envoy/blob/4b0b6540a1c34423813c6a81967cd6656c7b8f05/test/test_common/utility.h#L542.

**Comment by jplevyak** [comment]
Sorry for taking so long, I was preparing for Kubecon.  I'll be back on this ASAP.

File: test/mocks/grpc/mocks.h (hunk 22 above)
Comment (jplevyak) [reply test/mocks/grpc/mocks.h]: Checked the result of Parse.  PTAL

File: test/common/config/delta_subscription_test_harness.h (hunk 21 above)
Comment (jplevyak) [reply test/common/config/delta_subscription_test_harness.h]: done.

File: source/common/grpc/typed_async_client.h (hunk 20 above)
Comment (jplevyak) [reply source/common/grpc/typed_async_client.h]: Moved into Internal namespace.  PTAL

File: source/common/grpc/typed_async_client.h (hunk 19 above)
Comment (jplevyak) [reply source/common/grpc/typed_async_client.h]: done. AFAICT

File: source/common/grpc/async_client_impl.cc (hunk 18 above)
Comment (jplevyak) [reply source/common/grpc/async_client_impl.cc]: no, reverted

File: source/common/grpc/BUILD (hunk 17 above)
Comment (jplevyak) [reply source/common/grpc/BUILD]: removed.

File: test/mocks/grpc/mocks.h (hunk 22 above)
Comment (htuch) [reply test/mocks/grpc/mocks.h]: Maybe rename to `ProtoInBuffer` or something, I keep reading this as `ProtobBufEq`, trying to avoid poison tipped needles.

**Comment by jplevyak** [comment]
The MacOS and coverage failures seem to be unrelated.  lizan do you have any more comments?
I am not authorized to merge so I will continue to monitor this thread.  Thanks for the reviews!

**Comment by htuch** [comment]
@jplevyak the coverage is below the threshold, and we're getting TSAN flakes. Can you merge master and try again, plus maybe look to see if you can fill in any trivial coverage stuff missing? This PR is good to go modulo this.

**Comment by jplevyak** [comment]
I think the coverage is in pretty good shape. If there is anything which stands out, please tell me.

File: source/common/grpc/common.h [hunk 23]
```diff
@@ -159,6 +160,11 @@ class Common {
    */
   static void prependGrpcFrameHeader(Buffer::Instance& buffer);
 
+  template <typename P> static bool parseBufferInstance(Buffer::InstancePtr&& buffer, P& proto) {
```
Comment (lizan) [inline source/common/grpc/common.h]: This doesn't have to be templatized? You should be able to use `Protobuf::Message& proto`.

Comment (jplevyak) [reply source/common/grpc/common.h]: Done.

File: source/common/grpc/google_async_client_impl.cc [hunk 24]
```diff
@@ -317,7 +317,8 @@ void GoogleAsyncStreamImpl::handleOpCompletion(GoogleAsyncTag::Operation op, boo
   }
   case GoogleAsyncTag::Operation::Read: {
     ASSERT(ok);
-    if (!callbacks_.onReceiveMessageRaw(GoogleGrpcUtils::makeBufferInstance(read_buf_))) {
+    auto buffer = (GoogleGrpcUtils::makeBufferInstance(read_buf_));
```
Comment (htuch) [inline source/common/grpc/google_async_client_impl.cc]: Nit: superfluous parens.

File: source/common/grpc/google_grpc_utils.cc [hunk 25]
```diff
@@ -78,7 +78,9 @@ Buffer::InstancePtr GoogleGrpcUtils::makeBufferInstance(const grpc::ByteBuffer&
   // NB: ByteBuffer::Dump moves the data out of the ByteBuffer so we need to ensure that the
   // lifetime of the Slice(s) exceeds our Buffer::Instance.
   std::vector<grpc::Slice> slices;
-  byte_buffer.Dump(&slices);
+  if (!byte_buffer.Dump(&slices).ok()) {
```
Comment (htuch) [inline source/common/grpc/google_grpc_utils.cc]: Is there a way to mock this for testing?

File: test/common/grpc/google_grpc_utils_test.cc [hunk 26]
```diff
@@ -45,7 +45,7 @@ TEST(GoogleGrpcUtilsTest, MakeByteBuffer1) {
   buffer->add("test", 4);
   auto byte_buffer = GoogleGrpcUtils::makeByteBuffer(std::move(buffer));
   std::vector<grpc::Slice> slices;
-  byte_buffer.Dump(&slices);
+  RELEASE_ASSERT(byte_buffer.Dump(&slices).ok(), "");
```
Comment (htuch) [inline test/common/grpc/google_grpc_utils_test.cc]: When can `Dump` fail? We should only use `RELEASE_ASSERT` in situations where it's effectively OOM and some unrecoverable situation exists. Otherwise, there should be error handling.

Comment (PiotrSikora) [reply test/common/grpc/google_grpc_utils_test.cc]: See discussion in #6917.

Note: `RELEASE_ASSERT` is used only in tests, and failure is handled properly in the real code.

Comment (htuch) [reply test/common/grpc/google_grpc_utils_test.cc]: This is fine in the tests then. I would still like to see the `Dump` condition mocked and tested if gRPC has mocks that can support this.

File: source/common/grpc/google_grpc_utils.cc (hunk 25 above)
Comment (jplevyak) [reply source/common/grpc/google_grpc_utils.cc]: Even the gRPC library provided by Google doesn't test for Dump failure.  Here is the only test:

TEST_F(ByteBufferTest, Dump) {                                                                                                                                                                                                                                                                                                                                                                                                            
  grpc_slice hello = grpc_slice_from_copied_string(kContent1);                                                                                                                                                                                                                                                                                                                                                                            
  grpc_slice world = grpc_slice_from_copied_string(kContent2);                                                                                                                                                                                                                                                                                                                                                                            
  std::vector<Slice> slices;                                                                                                                                                                                                                                                                                                                                                                                                              
  slices.push_back(Slice(hello, Slice::STEAL_REF));                                                                                                                                                                                                                                                                                                                                                                                       
  slices.push_back(Slice(world, Slice::STEAL_REF));                                                                                                                                                                                                                                                                                                                                                                                       
  ByteBuffer buffer(&slices[0], 2);                                                                                                                                                                                                                                                                                                                                                                                                       
  slices.clear();                                                                                                                                                                                                                                                                                                                                                                                                                         
  (void)buffer.Dump(&slices);                                                                                                                                                                                                                                                                                                                                                                                                             
  EXPECT_TRUE(SliceEqual(slices[0], hello));                                                                                                                                                                                                                                                                                                                                                                                              
  EXPECT_TRUE(SliceEqual(slices[1], world));                                                                                                                                                                                                                                                                                                                                                                                              
}

File: test/common/grpc/google_grpc_utils_test.cc (hunk 26 above)
Comment (jplevyak) [reply test/common/grpc/google_grpc_utils_test.cc]: This is a test, so it is a RELEASE__ASSERT.  It should never fail.   For the production code I am handling the error.

File: source/common/grpc/google_async_client_impl.cc (hunk 24 above)
Comment (jplevyak) [reply source/common/grpc/google_async_client_impl.cc]: Done.

File: test/common/grpc/google_grpc_utils_test.cc (hunk 26 above)
Comment (jplevyak) [reply test/common/grpc/google_grpc_utils_test.cc]: See above regarding testing Dump().

File: source/common/grpc/google_grpc_utils.cc (hunk 25 above)
Comment (htuch) [reply source/common/grpc/google_grpc_utils.cc]: Yeah, this is fine. I think if this was Envoy code we would want a test, but it's not worth the effort to deal with this in external deps that aren't written with mocking in mind.
